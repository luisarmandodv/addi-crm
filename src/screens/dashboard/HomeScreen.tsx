import React, { useEffect, useState } from 'react';
import {
  View,
  StyleSheet,
  StatusBar,
  ActivityIndicator,
  ScrollView,
  SafeAreaView,
  SectionList,
  TouchableOpacity,
} from 'react-native';
import { Avatar, Title, Caption, Text } from 'react-native-paper';
import { useNavigation } from '@react-navigation/native';
import { useGetUsers } from '@hooks/useGetUsers';
import Modal from '@components/Modal';
import useClientsStore from '@stores/useClientsStore';
import useModal from '@stores/useModal';
import { colors } from '@theme/colors';
import { formStyles, typography } from '@theme/globalStyles';
import { UserData } from 'src/types';

export function HomeScreen() {
  const setIsModalVisible = useModal(
    (state: { setIsModalVisible: (value: boolean) => void }) =>
      state.setIsModalVisible
  );
  const { leads, prospects } = useClientsStore(({ leads, prospects }) => ({
    leads,
    prospects,
  }));
  const { data, isLoading } = useGetUsers();
  const { navigate }: any = useNavigation();

  const Loader = React.useMemo(() => {
    return (
      <ActivityIndicator
        testID="loader"
        hidesWhenStopped
        size="large"
        animating={isLoading}
        style={styles.activityIndicator}
      />
    );
  }, [isLoading]);

  const list = React.useMemo(() => {
    const sectionList = []
    
    if (leads && leads.length) {
      sectionList.push({
        title: 'Leads',
        data: leads as UserData[],
      })
    }

    if (prospects && prospects.length) {
      sectionList.push({
        title: 'Prospect',
        data: prospects as UserData[],
      });
    }

    return sectionList;
  }, [leads, prospects]);

  const List = React.useMemo(() => {
    return (
      <SectionList
        sections={list}
        keyExtractor={(item, index) => item.firstName + index.toString()}
        renderSectionHeader={({ section: { title } }) => (
          <Title style={styles.header}>{title}</Title>
        )}
        ListEmptyComponent={<View><Text>Empty</Text></View>}
        renderItem={({ item, section, index }) => {
          return (
            <TouchableOpacity
              testID={`button`}
              onPress={() => {
                navigate('UserDetails', { user: item, section: section.title });
              }}
            >
              <View style={styles.itemContainer}>
                <View style={{ flexDirection: 'row', alignItems: 'center' }}>
                  <View style={{ marginHorizontal: 15 }}>
                    <Avatar.Image
                      style={{
                        height: 50,
                        width: 50,
                        borderRadius: 50 / 2,
                        alignSelf: 'center',
                      }}
                      source={{
                        uri: item.image,
                      }}
                      size={50}
                    />
                  </View>
                  <View>
                    <Title style={typography.paragraph}>
                      {item.firstName + ' ' + item.lastName}
                    </Title>
                    <Caption style={{ color: colors.gray }}>
                      {item.email}
                    </Caption>
                    {section.title === 'Leads' ? (
                      <Text style={{ color: colors.warning }}>In Progress</Text>
                    ) : (
                      <Text style={{ color: colors.green }}>Approved</Text>
                    )}
                  </View>
                </View>
              </View>
            </TouchableOpacity>
          );
        }}
      />
    );
  }, [list]);

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle="dark-content" />
      {Loader}
      <Modal onLeftTextPress={() => setIsModalVisible(false)} />
      <ScrollView>{List}</ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: colors.primary },
  activityIndicator: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  itemContainer: {
    backgroundColor: colors.brown,
    marginBottom: 10,
    padding: 10,
    margin: 8,
    borderRadius: 10,
  },
  header: {
    marginLeft: 10,
    fontSize: 20,
    fontWeight: 'bold',
  },
});
