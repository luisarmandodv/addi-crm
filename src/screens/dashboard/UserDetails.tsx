import React, { useCallback, useMemo } from 'react';
import {
  View,
  StyleSheet,
  SafeAreaView,
  ScrollView,
  ActivityIndicator,
} from 'react-native';
import { Avatar, Title, Caption, Text, Button } from 'react-native-paper';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import { useNavigation } from '@react-navigation/native';
import { colors } from '@theme/colors';
import { useGetNRI } from '@hooks/useGetNRI';
import { useJudicialRecords } from '@hooks/useJudicialRecords';
import { useProspectQualification } from '@hooks/useProspectQualification';
import { AuthNavigatorRoutesprops } from '@routes/DashboardRoutes';
import useClientsStore from '@stores/useClientsStore';
import {
  setNotificationHandler,
  scheduleNotificationAsync,
} from 'expo-notifications';
import { UserData } from 'src/types';

interface ProfileProps {
  user: UserData;
  section: string;
}

type Props = NativeStackScreenProps<AuthNavigatorRoutesprops>;

setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

function UserInfo({ user }: { user: UserData }) {
  return (
    <View style={[styles.userInfoSection, styles.userInfoContainer]}>
      <View style={styles.userInfoRow}>
        <Avatar.Image source={{ uri: user.image }} size={80} />
        <View style={styles.userInfoText}>
          <Title
            style={[styles.title, styles.userInfoTitle]}
          >{`${user.firstName} ${user.lastName}`}</Title>
          <Caption style={styles.caption}>{user.username}</Caption>
        </View>
      </View>
      <View style={[styles.userInfoSection, styles.userInfoDetails]}>
        <View style={styles.row}>
          <Icon name="map-marker-radius" color="#777777" size={20} />
          <Text style={styles.loadingText}>{user.address.city}</Text>
        </View>
        <View style={styles.row}>
          <Icon name="phone" color="#777777" size={20} />
          <Text style={styles.loadingText}>{user.phone}</Text>
        </View>
        <View style={styles.row}>
          <Icon name="email" color="#777777" size={20} />
          <Text style={styles.loadingText}>{user.email}</Text>
        </View>
      </View>
    </View>
  );
}

function InfoBox({
  title,
  isLoading,
  data,
  isLeadsUser,
}: {
  title: string;
  isLoading: boolean;
  data: any;
  isLeadsUser: boolean;
}) {
  const Result = useMemo(() => {
    if (isLoading) {
      return (
        <ActivityIndicator
          testID={`${title}-loader`}
          hidesWhenStopped
          size="small"
          animating={true}
        />
      );
    }
    if (data?.data?.meetsRequirements || !isLeadsUser) {
      return <Icon name="check" color={colors.green} size={40} />;
    }
    return <Icon name="close" color={colors.danger} size={40} />;
  }, [isLoading, data, isLeadsUser]);

  return (
    <View style={styles.infoBox}>
      <Caption style={styles.infoBoxText}>{title}</Caption>
      {Result}
    </View>
  );
}

function InternalQualification({ user }: { user: UserData }) {
  const qualification = useProspectQualification();
  const percentage =
    user.qualification ||
    qualification.data?.qualification as number
  const isPropspect = Boolean(percentage > 60);
  const color = isPropspect ? colors.green : colors.danger;
  const setProspects = useClientsStore((state) => state.setProspects);
  const navigation = useNavigation();
  const result = useMemo(() => {
    if (qualification.isFetching || qualification.isLoading) {
      return (
        <ActivityIndicator
          testID={`internal-qualification-loader`}
          hidesWhenStopped
          size="small"
          animating={true}
        />
      );
    }
    return (
      <Text
        testID="internal-qualification-text"
        style={[styles.internalQualificationText, { color }]}
      >
        {percentage}/100
      </Text>
    );
  }, [qualification, color, percentage]);

  const onPress = useCallback(() => {
    const schedulingOptions = {
      content: {
        title: 'New Prospect Added',
        body: ``,
        sound: true,
        color: 'blue',
      },
      trigger: {
        seconds: 1,
      },
    };

    scheduleNotificationAsync(schedulingOptions);

    setProspects({ ...user, qualification: percentage });
    navigation.goBack();
  }, [user, percentage]);

  return (
    <>
      <View style={styles.leadsContainer}>
        <Title
          testID="internal-qualification-title"
          style={[styles.title, styles.userInfoTitle]}
        >
          Internal Qualification
        </Title>
        {result}
      </View>
      {!user.qualification && <Button
        mode="contained"
        disabled={!isPropspect}
        onPress={onPress}
        style={{ marginTop: 50, marginHorizontal: 30 }}
      >
        Add new prospect
      </Button>}
    </>
  );
}

export function UserDetails({ route }: Props) {
  const { user, section } = route.params as ProfileProps;
  const isLeadsUser = Boolean(section === 'Leads');
  const NRI = useGetNRI(isLeadsUser);
  const juficialRecords = useJudicialRecords(isLeadsUser);

  const renderInternalQualification = useMemo(() => {
    if (NRI.isFetching && juficialRecords.isFetching) return;
    if (
      !isLeadsUser ||
      (NRI.data?.meetsRequirements && juficialRecords.data?.meetsRequirements)
    ) {
      return <InternalQualification user={user} />;
    }
  }, [NRI, juficialRecords, user, isLeadsUser]);

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        <UserInfo user={user} />
        <View style={styles.infoBoxWrapper}>
          <InfoBox
            title="National Registry Identification"
            isLoading={NRI.isLoading || NRI.isFetching}
            data={NRI}
            isLeadsUser={isLeadsUser}
          />
          <InfoBox
            title="Judicial Records"
            isLoading={juficialRecords.isLoading || juficialRecords.isFetching}
            data={juficialRecords}
            isLeadsUser={isLeadsUser}
          />
        </View>
        {renderInternalQualification}
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.primary,
  },
  userInfoSection: {
    paddingHorizontal: 30,
    marginBottom: 25,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
  },
  caption: {
    fontSize: 14,
    lineHeight: 14,
    fontWeight: '500',
  },
  row: {
    flexDirection: 'row',
    marginBottom: 16,
  },
  infoBoxWrapper: {
    margin: 10,
    padding: 10,
    backgroundColor: colors.brown,
    borderRadius: 10,
    flexDirection: 'row',
    height: 100,
  },
  infoBox: {
    width: '50%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  infoBoxText: {
    marginBottom: 10,
  },
  loadingText: {
    color: '#777777',
  },
  internalQualificationText: {
    fontSize: 25,
    alignSelf: 'center',
    marginTop: 10,
  },
  userInfoContainer: {
    backgroundColor: colors.brown,
    marginBottom: 10,
    padding: 10,
    margin: 8,
    borderRadius: 10,
  },
  leadsContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    padding: 10,
    borderRadius: 10,
    backgroundColor: colors.brown,
  },
  userInfoRow: { flexDirection: 'row', marginLeft: 7 },
  userInfoText: { marginLeft: 20 },
  userInfoTitle: { marginTop: 15, marginBottom: 5 },
  userInfoDetails: { marginTop: '10%', marginLeft: '-10%' },
});
