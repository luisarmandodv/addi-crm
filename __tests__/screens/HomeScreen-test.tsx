import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react-native';
import { HomeScreen } from '@screens/dashboard/HomeScreen';
import { QueryClient, QueryClientProvider } from "react-query";

jest.mock("axios");

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

const fakeData = {
  users: [
    { firstName: 'John', lastName: 'Doe', email: 'john@example.com', image: 'john.jpg' },
    // Add more fake user data as needed
  ],
};

jest.mock('@hooks/useGetUsers', () => ({
  useGetUsers: () => ({
    data: fakeData,
  }),
}));

describe('HomeScreen component', () => {
  const client = new QueryClient();

  it('Should render correctly component', () => {
    render(
      <QueryClientProvider client={client}>
        <HomeScreen />
      </QueryClientProvider>
    );
  });

  it('Should render snapshot', () => {
    // expect(tree).toMatchSnapshot();
    render(
      <QueryClientProvider client={client}>
        <HomeScreen />
      </QueryClientProvider>
    );


    expect(screen.toJSON()).toMatchSnapshot();
  });
  it('renders loading indicator when isLoading is true', () => {
    const { getByTestId } = render(
      <QueryClientProvider client={client}>
        <HomeScreen />
      </QueryClientProvider>);
    const loader = getByTestId('loader');
    expect(loader).toBeTruthy();
  });

  it('renders sections and items when data is available', async () => {

    const { findByText } = render(
      <QueryClientProvider client={client}>
        <HomeScreen />
      </QueryClientProvider>
    );
    await waitFor(() => {
      expect(findByText('Leads')).toBeTruthy();
      expect(findByText('John Doe')).toBeTruthy();
      expect(findByText('john@example.com')).toBeTruthy();
    });
  });

  it('calls navigate when an item is pressed', () => {
    const { getByTestId } = render(
      <QueryClientProvider client={client}>
        <HomeScreen />
      </QueryClientProvider>);
    const item = getByTestId('button'); // Replace with the name of an actual item
    fireEvent.press(item);

    expect(mockedNavigate).toHaveBeenCalledWith('UserDetails', { user: expect.any(Object), section: 'Leads' });
  });
});
