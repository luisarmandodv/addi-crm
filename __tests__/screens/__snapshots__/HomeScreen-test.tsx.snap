// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen component Should render snapshot 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#82ccdd",
      "flex": 1,
    }
  }
>
  <ActivityIndicator
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="loader"
  />
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    presentationStyle="pageSheet"
    transparent={true}
    visible={false}
  />
  <RCTScrollView>
    <View>
      <RCTScrollView
        data={
          Array [
            Object {
              "data": Array [
                Object {
                  "email": "john@example.com",
                  "firstName": "John",
                  "image": "john.jpg",
                  "lastName": "Doe",
                },
              ],
              "title": "Leads",
            },
            Object {
              "data": Array [],
              "title": "Prospects",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={
          Array [
            0,
            3,
          ]
        }
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(28, 27, 31, 0.87)",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 20,
                        "letterSpacing": 0.15,
                        "lineHeight": 30,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 20,
                        "fontWeight": "bold",
                        "marginLeft": 10,
                      },
                    ],
                  ],
                ]
              }
            >
              Leads
            </Text>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="button"
            >
              <View
                style={
                  Object {
                    "backgroundColor": "#f0ece2",
                    "borderRadius": 10,
                    "margin": 8,
                    "marginBottom": 10,
                    "padding": 10,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "marginHorizontal": 15,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "backgroundColor": "rgba(103, 80, 164, 1)",
                            "borderRadius": 25,
                            "height": 50,
                            "width": 50,
                          },
                          Object {
                            "alignSelf": "center",
                            "borderRadius": 25,
                            "height": 50,
                            "width": 50,
                          },
                        ]
                      }
                    >
                      <Image
                        accessibilityIgnoresInvertColors={true}
                        source={
                          Object {
                            "uri": "john.jpg",
                          }
                        }
                        style={
                          Object {
                            "borderRadius": 25,
                            "height": 50,
                            "width": 50,
                          }
                        }
                      />
                    </View>
                  </View>
                  <View>
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "rgba(28, 27, 31, 1)",
                          },
                          Object {
                            "textAlign": "left",
                          },
                          Array [
                            Object {
                              "textAlign": "left",
                            },
                            Object {
                              "color": "rgba(28, 27, 31, 0.87)",
                              "writingDirection": "ltr",
                            },
                            Array [
                              Object {
                                "fontSize": 20,
                                "letterSpacing": 0.15,
                                "lineHeight": 30,
                                "marginVertical": 2,
                              },
                              Object {
                                "color": "#3c6382",
                                "fontSize": 16,
                                "marginTop": 8,
                              },
                            ],
                          ],
                        ]
                      }
                    >
                      John Doe
                    </Text>
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "rgba(28, 27, 31, 1)",
                          },
                          Object {
                            "textAlign": "left",
                          },
                          Array [
                            Object {
                              "textAlign": "left",
                            },
                            Object {
                              "color": "rgba(28, 27, 31, 0.54)",
                              "writingDirection": "ltr",
                            },
                            Array [
                              Object {
                                "fontSize": 12,
                                "letterSpacing": 0.4,
                                "lineHeight": 20,
                                "marginVertical": 2,
                              },
                              Object {
                                "color": "#898A8D",
                              },
                            ],
                          ],
                        ]
                      }
                    >
                      john@example.com
                    </Text>
                    <Text
                      style={
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(28, 27, 31, 1)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "letterSpacing": 0,
                          },
                          Object {
                            "writingDirection": "ltr",
                          },
                          Object {
                            "color": "#e58e26",
                          },
                        ]
                      }
                    >
                      In Progress
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          />
          <View
            onLayout={[Function]}
            style={null}
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(28, 27, 31, 0.87)",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 20,
                        "letterSpacing": 0.15,
                        "lineHeight": 30,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 20,
                        "fontWeight": "bold",
                        "marginLeft": 10,
                      },
                    ],
                  ],
                ]
              }
            >
              Prospects
            </Text>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          />
        </View>
      </RCTScrollView>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
